sycl/win_proxy_loader/win_proxy_loader.cpp

```diff
@@ -0,0 +1,186 @@
+  assert(Ret > 0 && "GetModuleFileNameA failed");
+  (void)Ret;
+
+  BOOL RetCode = PathRemoveFileSpecA(reinterpret_cast<LPSTR>(&Path));
+  assert(RetCode && "PathRemoveFileSpecA failed");
+  (void)RetCode;
+
+  return Path;
+}
+
+// these are cribbed from include/sycl/detail/pi.hpp
+// a new plugin must be added to both places.
+#define __SYCL_OPENCL_PLUGIN_NAME "pi_opencl.dll"
+#define __SYCL_LEVEL_ZERO_PLUGIN_NAME "pi_level_zero.dll"
+#define __SYCL_CUDA_PLUGIN_NAME "pi_cuda.dll"
+#define __SYCL_ESIMD_EMULATOR_PLUGIN_NAME "pi_esimd_emulator.dll"
+#define __SYCL_HIP_PLUGIN_NAME "libpi_hip.dll"
+#define __SYCL_UNIFIED_RUNTIME_PLUGIN_NAME "pi_unified_runtime.dll"
+
+// ------------------------------------
+
+static std::map<std::string, void *> dllMap;
+
+/// load the five libraries and store them in a map.
+void preloadLibraries() {
+  // Suppress system errors.
+  // Tells the system to not display the critical-error-handler message box.
+  // Instead, the system sends the error to the calling process.
+  // This is crucial for graceful handling of plugins that couldn't be
+  // loaded, e.g. due to missing native run-times.
+  // Sometimes affects L0 or the unified runtime.
+  // TODO: add reporting in case of an error.
+  // NOTE: we restore the old mode to not affect user app behavior.
+  //
+  UINT SavedMode = SetErrorMode(SEM_FAILCRITICALERRORS);
+  // Exclude current directory from DLL search path
+  if (!SetDllDirectoryA("")) {
+    assert(false && "Failed to update DLL search path");
+  }
+
+  // this path duplicates sycl/detail/pi.cpp:initializePlugins
+  const std::string LibSYCLDir = getCurrentDSODir() + DirSep;
+
+  std::string ocl_path = LibSYCLDir + __SYCL_OPENCL_PLUGIN_NAME;
+  dllMap.emplace(ocl_path, LoadLibraryA(ocl_path.c_str()));
+
+  std::string l0_path = LibSYCLDir + __SYCL_LEVEL_ZERO_PLUGIN_NAME;
+  dllMap.emplace(l0_path, LoadLibraryA(l0_path.c_str()));
+
+  std::string cuda_path = LibSYCLDir + __SYCL_CUDA_PLUGIN_NAME;
+  dllMap.emplace(cuda_path, LoadLibraryA(cuda_path.c_str()));
+
+  std::string esimd_path = LibSYCLDir + __SYCL_ESIMD_EMULATOR_PLUGIN_NAME;
+  dllMap.emplace(esimd_path, LoadLibraryA(esimd_path.c_str()));
+
+  std::string hip_path = LibSYCLDir + __SYCL_HIP_PLUGIN_NAME;
+  dllMap.emplace(hip_path, LoadLibraryA(hip_path.c_str()));
+
+  std::string ur_path = LibSYCLDir + __SYCL_UNIFIED_RUNTIME_PLUGIN_NAME;
+  dllMap.emplace(ur_path, LoadLibraryA(ur_path.c_str()));
+
+  // Restore system error handling.
+  (void)SetErrorMode(SavedMode);
+  if (!SetDllDirectoryA(nullptr)) {
+    assert(false && "Failed to restore DLL search path");
+  }
+}
+
+/// windows_pi.cpp:loadOsLibrary() calls this to get the DLL we loaded earlier.
+__declspec(dllexport) void *getPreloadedPlugin(const std::string &PluginPath) {
+
+  auto match = dllMap.find(
+      PluginPath); // result might be nullptr, which is perfectly valid.
+  if (match == dllMap.end()) {
+    // unit testing? Just return nullptr.
```

Comment:
Could you please improve the comment? It's not very clear why we need to return `nullptr` here.

Reply:
done